precompiles/utils/macro/src/precompile/parse.rs

```diff
@@ -0,0 +1,490 @@
+		let attrs = attr::take_attributes::<attr::MethodAttr>(&mut method.attrs)?;
+
+		// If there are no attributes it is a private function and we ignore it.
+		if attrs.is_empty() {
+			return Ok(());
+		}
+
+		// A method cannot have modifiers if it isn't a fallback and/or doesn't have a selector.
+		let mut used = false;
+
+		let method_name = method.sig.ident.clone();
+		let mut modifier = Modifier::NonPayable;
+		let mut solidity_arguments_type: Option<String> = None;
+		let mut arguments = vec![];
+		let mut is_fallback = false;
+		let mut selectors = vec![];
+		let initial_arguments = if self.tagged_as_precompile_set { 2 } else { 1 };
+
+		if let Some(attr::MethodAttr::Discriminant(span)) = attrs.first() {
+			let span = *span;
+			if !self.tagged_as_precompile_set {
+				let msg = "The impl block must be tagged with `#[precompile::precompile_set]` for
+				the discriminant attribute to be used";
+				return Err(syn::Error::new(span, msg));
+			}
+
+			if self.precompile_set_discriminant_fn.is_some() {
+				let msg = "A PrecompileSet can only have 1 discriminant function";
+				return Err(syn::Error::new(span, msg));
+			}
+
+			if attrs.len() != 1 {
+				let msg = "The discriminant attribute must be the only precompile attribute of \
+				a function";
+				return Err(syn::Error::new(span, msg));
+			}
+
+			let span = method.sig.span();
+
+			if method.sig.inputs.len() != 1 {
+				let msg = "The discriminant function must only take the code address (H160) as \
+				parameter.";
+				return Err(syn::Error::new(span, msg));
+			}
+
+			let msg = "The discriminant function must return an Option<_> (no type alias)";
+
+			let return_type = match &method.sig.output {
+				syn::ReturnType::Type(_, t) => t.as_ref(),
+				_ => return Err(syn::Error::new(span, msg)),
+			};
+
+			let return_path = match return_type {
+				syn::Type::Path(p) => p,
+				_ => return Err(syn::Error::new(span, msg)),
+			};
+
+			if return_path.qself.is_some() {
+				return Err(syn::Error::new(span, msg));
+			}
+
+			let return_path = &return_path.path;
+
+			if return_path.leading_colon.is_some() || return_path.segments.len() != 1 {
+				return Err(syn::Error::new(span, msg));
+			}
+
+			let return_segment = &return_path.segments[0];
+
+			if return_segment.ident.to_string() != "Option" {
+				return Err(syn::Error::new(return_segment.ident.span(), msg));
+			}
+
+			let option_arguments = match &return_segment.arguments {
+				syn::PathArguments::AngleBracketed(args) => args,
+				_ => return Err(syn::Error::new(return_segment.ident.span(), msg)),
+			};
+
+			if option_arguments.args.len() != 1 {
+				let msg = "Option type should only have 1 type argument";
+				return Err(syn::Error::new(option_arguments.args.span(), msg));
+			}
+
+			let discriminant_type: &syn::Type = match &option_arguments.args[0] {
+				syn::GenericArgument::Type(t) => t,
+				_ => return Err(syn::Error::new(option_arguments.args.span(), msg)),
+			};
```

Comment:
Can you move this block in a new function (named "parse_discriminant_fn" for example) for better readability?